### PR TITLE
Relay ancestors of transactions

### PR DESCRIPTION
**Pull request by sdaftuar**
Respond to getdata requests for transactions that are ancestors of newly-announced transactions.  Currently, our software will request the parents of an orphan transaction from the peer who announced the orphan, but if our peer is running a recent Bitcoin Core version, the request would typically be dropped.  So this should improve propagation for transaction chains, particularly for nodes that are just starting up.

(continued from #14220, which is now just a bugfix PR)

Depends on:
 - [ ] #14220

### CODE REVIEW

**Review by jamesob — APPROVED** [review]
utACK https://github.com/bitcoin/bitcoin/pull/14318/commits/6b661802eb2baa206b3ba0f192be39c2b304cc00

For future reviewers: the INV for missing parents (which we currently expect may not be answered if relay interval on the remote node has lapsed) happens here: https://github.com/bitcoin/bitcoin/blob/d799efe21432ad55c41d5315f24c002bc8b3d119/src/net_processing.cpp#L2303-L2309

This PR is dependent on https://github.com/bitcoin/bitcoin/pull/14220 and probably shouldn't be merged before that.

File: test/functional/test_framework/messages.py [hunk 1]
```diff
@@ -1160,6 +1160,26 @@ def serialize(self):
...
+class msg_notfound():
+    command = b"notfound"
+
+    def __init__(self, vec=None):
+        if vec is None:
+            self.vec = []
+        else:
+            self.vec = vec
+
+    def deserialize(self, f):
```
Comment (practicalswift) [inline test/functional/test_framework/messages.py]: Could use a better variable name than `f`? :-)

Comment (jamesob) [reply test/functional/test_framework/messages.py]: This maintains consistency with existing methods in similar classes (and it's a one-line usage: not considerably different from, say, a list comprehension) so I think this is fine.

File: test/functional/test_framework/mininode.py [hunk 2]
```diff
@@ -295,6 +327,7 @@ def on_getdata(self, message): pass
     def on_getheaders(self, message): pass
     def on_headers(self, message): pass
     def on_mempool(self, message): pass
+    def on_notfound(self, message): pass
```
[the remark this answers is not in the document]
Comment (jamesob) [reply test/functional/test_framework/mininode.py]: This is required AFAICT due to https://github.com/bitcoin/bitcoin/pull/14318/files/6b661802eb2baa206b3ba0f192be39c2b304cc00#diff-a77567bbc57dc1bbc033560e5cdd3b7dR304.

[the remark this answers is not in the document]
Comment (practicalswift) [reply test/functional/test_framework/mininode.py]: Ah, makes sense!

File: test/functional/test_framework/messages.py (hunk 1 above)
Comment (practicalswift) [reply test/functional/test_framework/messages.py]: Now I see that `f` is used consistently. Makes sense to follow that.

**Comment by laanwj** [comment]
@sdaftuar do you have any specific reason for closing?

**Comment by sdaftuar** [comment]
@laanwj This builds on a PR that I closed because it needed rework, so I just didn't want to leave this open when it wasn't ready for review.

**Comment by laanwj** [comment]
@sdaftuar thanks for the explanation, makes sense

**Pull request closed without merging**